Fix the @since attribute for the Uchar.hash function

shindere opened the pull request:
This function was documented as having been introduced in OCaml 5.3
whereas it is present since the initial implementation of the
Uchar module in 940144f30487ead603d51ad4783542fe5410b821, which was part
of OCaml 4.03 released in April 2016.

(Should be backported where appropriate)

kit-ty-kate commented on stdlib/uchar.mli:
```diff
@@ -107,7 +107,7 @@ val hash : t -> int
     {!Hashtbl.Make}.
 
-    @since 5.3 *)
+    @since 4.03 *)
```
Since the module itself was introduced in 4.03 i believe the usual etiquette is to not have a `@since` annotation when the function was introduced along with the module.
```suggestion
    {!Hashtbl.Make}. *)
```

shindere commented:
Ah. Just for sake of completeness this anotation was added in #13240.

shindere commented:
Many thanks for your prompt and helpful feedback, @kit-ty-kate.

I have just applied your suggestion.

nojb reviewed (commented):
The function existed since much longer, but its semantics changed in 5.3. It seems like a good idea to mention this in the docstring.

shindere commented:
So you mean to just close this PR, or rather to mention the semantic
change but not necessarily to restore the @since attribute?

nojb commented:
> So you mean to just close this PR, or rather to mention the semantic change but not necessarily to restore the @SInCE attribute?

Yes, mention the semantic change and remove the `@since` attribute, which is inaccurate.

hyphenrf commented:
I was responsible for this doc comment, that's my mistake :pray:.
Coincidentally I noticed this a while back and wrote a patch for it, but got distracted with whether or not I should use crossrefs and then forgot to follow up with the PR:
```patch
; git diff -R hazem/fix-uchar-hash-tag stdlib/uchar.mli
diff --git b/stdlib/uchar.mli a/stdlib/uchar.mli
index 155d8f8..9bd9272 100644
--- b/stdlib/uchar.mli
+++ a/stdlib/uchar.mli
@@ -107,7 +107,8 @@ val hash : t -> int
     This function allows this module to be passed as an argument to the functor
     {!Hashtbl.Make}.
 
-    @since 5.3 *)
+    @before 5.3 The hashing algorithm was different. Use [Hashtbl.rebuild] for
+    stored tables which used this hashing function *)
 
 (** {1:utf UTF codecs tools}
```
I propose using `@before` like this. Seems like the appropriate use of it.

nojb commented:
> I propose using `@before` like this. Seems like the appropriate use of it.

Indeed, good suggestion. Thanks!

shindere commented:
Haz (2025/03/27 06:28 -0700):
> hyphenrf left a comment (ocaml/ocaml#13892)
>
> Coincidentally I noticed this the other day and wrote a diff for it:
> ```patch
> ; git diff hazem/fix-uchar-hash-tag stdlib/uchar.mli
> diff --git a/stdlib/uchar.mli b/stdlib/uchar.mli
> index 9bd9272..155d8f8 100644
> --- a/stdlib/uchar.mli
> +++ b/stdlib/uchar.mli
> @@ -107,8 +107,7 @@ val hash : t -> int
>      This function allows this module to be passed as an argument to the functor
>      {!Hashtbl.Make}.
>
> -    @since 5.3 *)
> +    @before 5.3 The hashing algorithm was different. Use [Hashtbl.rebuild] for
> +    stored tables which used this hashing function *)
>
>  (** {1:utf UTF codecs tools}
> ```

Thanks!

What does `(1)` refer to?

hyphenrf commented:
No problem, all

> What does `(1)` refer to?

I'm unsure. I couldn't locate a literal `(1)` in my diff.. only the utf header `{1:utf ...}` from the diff context showing an extra part of the docs but I don't think that's what you meant

nojb commented:
@shindere can you update the PR as suggested by @hyphenrf? We can then merge. Thanks!

hyphenrf commented on stdlib/uchar.mli:
```diff
@@ -105,9 +105,7 @@ val seeded_hash : int -> t -> int
-    {!Hashtbl.Make}.
-
-    @since 5.3 *)
+    {!Hashtbl.Make}. *)
```
I made the suggestion using the Github UI for convenience
```suggestion
    @before 5.3 The hashing algorithm was different. Use [Hashtbl.rebuild] for
    stored tables which used this hashing function *)
```

shindere committed:
Get rid of the @since attribute for the Uchar.hash function

Given that this function is present in the module since it was
introduced, the convension is to not have any @since attribute
at the function level, since the module-level one applies.

This commit reverts the addition of the attribute done in PR #13240, see
commit 9585cfe264c4d9b4185f60ebfbbec3f8e54d1432.

shindere commented:
@nojb done, sorry for the delay, was unwell.

shindere commented:
Thanks, @hyphenrf. I did integrate your diff (with a slightly different
fomatting) in the latest version of the PR. I cannot find what I was
refering too with the `(1)` thing. :(

nojb reviewed (approved):
LGTM
